httpclient/src/main/java/com/backblaze/b2/client/webApiHttpClient/B2WebApiHttpClientImpl.java

```diff
@@ -124,6 +129,46 @@ public void getContent(String url,
         }
     }
 
+    /**
+     * HEADSs to a web service that returns content, and returns the headers.
+     *
+     * @param url the url to head to
+     * @param headersOrNull the headers, if any.
+     * @return the headers of the response.
+     * @throws B2Exception if there's any trouble
+     */
+    @Override
+    public Map<String, String> head(String url, B2Headers headersOrNull)
+            throws B2Exception {
+
+        CloseableHttpResponse response = null;
+        try {
+            HttpHead head = new HttpHead(url);
+            if (headersOrNull != null) {
+                head.setHeaders(makeHeaders(headersOrNull));
+            }
+
+            response = clientFactory.create().execute(head);
+
+            int statusCode = response.getStatusLine().getStatusCode();
+            if (statusCode == HttpStatus.SC_OK) {
+                return Arrays.stream(response.getAllHeaders())
+                        .collect(Collectors.toMap(Header::getName, Header::getValue));
+            } else {
+                if (statusCode == HttpStatus.SC_NOT_FOUND) {
```

Comment:
can we use code similar to what we do for other http methods that return non-200 status codes?
in particular, can we use?

                throw extractExceptionFromErrorResponse(response, responseText);

I think your goal here was to distinguish the "not found" case.  Calling extractExceptionFromErrorResponse() will use B2Exception.create() which picks an appropriate B2Exception subclass based on the http status code.  In particular, for 404, it should return a B2NotFoundException.  will that do what you want?

Reply:
that is an *excellent* point.  :)
thinking....

Reply:
i still want to create the right B2Exception subclass.
we have access to the http request's status code.
still thinking about the other arguments.

Reply:
I currently think that the right thing is the following:

   throw B2Exception.create(null, statusCode, getRetryAfterSecondsOrNull(response), "");

PLUS teaching the built-in exception classes to provide a reasonable code if they're constructed with code=null.  Let me see about adding those defaults so that you can take advantage of them.